### PR TITLE
X509Certificate is now immutable to make it consistent with CoreFX.

### PR DESCRIPTION
**Pull request by baulig**
`X509Certificate` is now immutable to make it consistent with CoreFX.

* All `X509Certificate.Import()` methods now `PlatformNotSupportedException (SR.NotSupported_ImmutableX509Certificate)`.

* The `X509Certificate2.PrivateKey` setter unconditionally throws `PlatformNotSupportedException ()` on all platforms.

* All `X509Certificate` and `X509Certificate2` constructors, as well as `X509Certificate2Collection` and `X509Store` have been fixed to not call `Import()` anymore.

* The following tests have been removed:
   - In `X509Cert20Test.cs`: `GetObjectData_Null()`, `GetObjectData()`, `Ctor_Serialization_Null()`, `Ctor_Serialization()` and `Pkcs7_Import()`; `Empty()` has been kept, but patched to no longer call `Import()` at the end.
  - In `X509Certificate2Test`: `Pkcs7_Import()`, `Empty_PrivateKey_Set()`

* `mcs/tools/linker/Tests/System/test-security.cs`: Fixed the test to use the `X509Certificate2(byte[])` constructor instead of calling `Import()`.

### CODE REVIEW

File: mcs/class/System/System.Security.Cryptography.X509Certificates/X509Certificate2.cs [hunk 1]
```diff
@@ -216,46 +218,32 @@ public string GetNameInfo (X509NameType nameType, bool forIssuer)
 
 		public override void Import (byte[] rawData) 
 		{
-			Import (rawData, (string)null, X509KeyStorageFlags.DefaultKeySet);
+			base.Import (rawData);
 		}
```
Comment (filipnavara) [inline mcs/class/System/System.Security.Cryptography.X509Certificates/X509Certificate2.cs]: No need to override at all, same with the methods below.

File: mcs/class/System/Test/System.Security.Cryptography.X509Certificates/X509Certificate2Test.cs [hunk 2]
```diff
@@ -1619,13 +1611,6 @@ public void Empty_PrivateKey_Get ()
 			Assert.IsNull (new X509Certificate2 ().PrivateKey);
 		}
 
-		[Test]
-		[ExpectedException (typeof (CryptographicException))]
-		public void Empty_PrivateKey_Set ()
-		{
-			new X509Certificate2 ().PrivateKey = null;
-		}
```
Comment (filipnavara) [inline mcs/class/System/Test/System.Security.Cryptography.X509Certificates/X509Certificate2Test.cs]: Maybe keep this test and adjust the expected exception?

File: mcs/class/System/System.Security.Cryptography.X509Certificates/X509Certificate2.cs (hunk 1 above)
Comment (baulig) [reply mcs/class/System/System.Security.Cryptography.X509Certificates/X509Certificate2.cs]: I verbosely copied these from CoreFX's version.

**Comment by baulig** [comment]
Formerly known as "Duck Blue".  This is now rebased on recent master and ready for review and landing.

**Comment by marek-safar** [comment]
This breaks linker test

**Comment by baulig** [comment]
Which linker test?

**Comment by marek-safar** [comment]
https://jenkins.mono-project.com/job/test-mono-pull-request-amd64/16665/parsed_console/log_content.html#WARNING1

```
System.PlatformNotSupportedException: Cannot get `ISystemDependencyProvider`.
  at Mono.DependencyInjector.get_SystemProvider () [0x0003d] in /mnt/jenkins/workspace/test-mono-pull-request-amd64/mcs/class/corlib/Mono/DependencyInjector.cs:54 
  at System.Security.Cryptography.X509Certificates.X509Helper.get_CertificateProvider () [0x00000] in /mnt/jenkins/workspace/test-mono-pull-request-amd64/mcs/class/corlib/System.Security.Cryptography.X509Certificates/X509Helper.cs:45
```

**Comment by baulig** [comment]
@monojenkins build OS X x64

**Comment by baulig** [comment]
@monojenkins build Windows i386

**Comment by baulig** [comment]
@monojenkins build Windows x64

**Comment by baulig** [comment]
@marek-safar Looks like the linker error did not happen again after rebuilding.  It is in any case completely unrelated to this PR.  However, we need to keep an eye of this and if it happens again, I'll have to fix something.

@akoeplinger Is there any difference in configuration between the `OS X i386` and `OS X x64` lanes (apart from architecture)?  This linker error should either happen on all platforms or on none of them.

**Comment by baulig** [comment]
This linker failure is completely unrelated to the PR - I don't know what has caused that, but if the linker somehow fails on the dependency injector, that would affect all PRs on all platforms since we landed the DependencyInjector.

I am going to trigger a full rebuild on this and if this comes up again without linker failures, then can we please get this in on Monday, so I can move on and land "Ultramarine"?

**Comment by baulig** [comment]
@monojenkins csrpoj

**Comment by baulig** [comment]
@monojenkins csproj

**Comment by baulig** [comment]
@monojenkins build

**Comment by marek-safar** [comment]
build

**Comment by marek-safar** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
@monojenkins commit apidiff

**Pull request merged**